### DISCUSSION TITLE
How to fetch block content using IApi and custom page types

### DISCUSSION THREAD

**Discussion by gmullernh** [PiranhaCMS/piranha.core#1676] [Q&A] [answered]
Hello!
I'm using the `IApi` at a controller to fetch some custom pages types. 
The page model is using the `[PageType(Title = "MyPageType")]` attribute.
I'm serializing it using `Microsoft.AspNetCore.Mvc.Json` as a return from the action.

```CSharp
      List<MyPageType> pages = new List<MyPageType>();

      foreach (Guid pageId in _db.Pages
          .Where(p => p.Published <= DateTime.Now && p.PageTypeId == "MyTypePage")
          .Skip(pagenumber * amount)
          .Take(amount)
          .Select(p => p.Id)
          .ToList())
      {
          pages.Add(await _api.Pages.GetByIdAsync<MyTypePage>(pageId).ConfigureAwait(false));
      }

      return Json(pages);
```

Everything is working fine, I'm getting all the page's infos, except the blocks list, which is coming without the contents in the JSON response, like the example below. The block contains some content if I open it on the manager.

```JSON
    "blocks": [
      {
        "id": "c178679a-bc32-420c-bd85-5974d1b43b81",
        "type": "Piranha.Extend.Blocks.HtmlBlock"
      }
    ],
```

I've tried using the IModelLoader to see if it would recover the contents, but without success (the result is the same if using the IApi to get the content). I'm using this application as an API for consuming the CMS content (without the manager).

In another project, which uses the Manager, the StandardPage and the IModelLoader, the JSON response contains all blocks content.

```CSharp
      [Route("page")]
      public async Task<IActionResult> Page(Guid id, bool draft = false)
      {
          try
          {
              var model = await _loader.GetPageAsync<StandardPage>(id, HttpContext.User, draft);
              return Json(model);
          }
          catch (UnauthorizedAccessException)
          {
              return Unauthorized();
          }
      }
```

```JSON
    "blocks": [
        {
            "$type": "Piranha.Extend.Blocks.TextBlock, Piranha",
            "body": {
                "value": "Hello!"
            },
            "id": "50e72755-b645-4f55-baf7-105a2ca9c168",
            "type": "Piranha.Extend.Blocks.TextBlock"
        },
    ]
```

I thought it could be a question with the Microsoft's Json serializer, but it's working fine on the other project.
Might this be an issue of not having the manager or with the serialization of the custom model?

**Comment by tidyui** [comment, marked as answer]
You can’t use Microsoft’s JSON serializer with Piranha as it does not support serialization or deserialization of non finite types. Instead use Newtonsoft for serialization.

**Reply by gmullernh** [reply]
Thank you!
Using `Newtonsoft.Json.JsonConvert.SerializeObject` to serialize the type solved the issue.
